feat: add rate limit and faster default speed

wakamex opened the pull request:
### What I did
add rate limit, implementing similar logic to [alchemy-web3](https://github.com/alchemyplatform/alchemy-web3/blob/f0df9fbbfb2d1e2922b9ccaaabd6f0fe6d06af25/src/web3-adapter/sendJsonRpcPayload.ts) but with exponential falloff which the [alchemy docs](https://docs.alchemy.com/reference/throughput) tell me is preferred

<!-- The `fixes:` field denotes an issue that will be marked resolved by merging this PR -->
fixed: #60
also fixes [ape issue #613](https://github.com/ApeWorX/ape/issues/613) which suggests adding a rate limit `as another mix-in for ProviderAPI`. not sure what that means, but this modifies a sub-class of a sub-class of ProviderAPI.

### How I did it
for loop around _make_request call, if a call is successful it returns it, otherwise keeps going. if the for loop ends, it raises an error saying it's rate limited.

### How to verify it
I only checked this against an intro notebook called [hello_ape.ipynb](https://github.com/fmrmf/hello-ape/blob/main/console/hello-ape.ipynb) but i don't see why it shouldn't work generally.
In the two timed queries this is 6x (1.98s/0.335s) and 13x (202s/15.9s) faster.

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

fubuloubu reviewed (approved):
absolutely sexy!

@unparalleled-js would we want to upstream this as a mix-in in ape core per https://github.com/ApeWorX/ape/issues/613?

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -129,22 +136,37 @@ def get_virtual_machine_error(self, exception: Exception) -> VirtualMachineError
+                    if isinstance(error_data, dict)
+                    else error_data
+                )
+                if message.__contains__("exceeded its compute units"):
```
```suggestion
                if "exceeded its compute units" in message:
```

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -26,6 +27,12 @@ class Alchemy(Web3Provider, UpstreamProvider):
+        super().__init__(**kwargs)
+        self.block_page_size = 5000
+        self.concurrency = 1
+        print(f"setting block_page_size to {self.block_page_size} and concurrency to {self.concurrency}")
```
- [ ] Replace `print()` statements with the logger.

```python
from ape.logging import logger

logger.debug()
# or
logger.info()
# etc.
```

wakamex replied:
removed

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -129,22 +136,37 @@ def get_virtual_machine_error(self, exception: Exception) -> VirtualMachineError
-                else AlchemyProviderError
-            )
-            raise cls(message) from err
+        MIN_RETRY_DELAY = 1000
```
In Python, constants are typically capitalized only when defined at the global scope.
It may make more sense to make all of these function arguments on `_make_request()` so that hackers can change their values if they want to get all internal on us.

wakamex replied:
converted to lower case and put as default callable parameters

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -129,22 +136,37 @@ def get_virtual_machine_error(self, exception: Exception) -> VirtualMachineError
+        RETRY_BACKOFF_FACTOR = 2
+        MAX_RETRY_DELAY = 30000
+        MAX_RETRIES = 3
+        # RETRY_INTERVAL = 1000
```
Can this be deleted?

```suggestion
```

unparalleled-js reviewed (commented):
Amazing! I want, I want.

Regarding the mixin: I don't think making a mixin out of this would be an easy task because:

1. The error message for exceeding computational units may differ from provider to provider
2. We need to retain the Alchemy-specific error handling we do here already, such as the 403-like handling
3. It increases the complexity of `super()` and base-classes, causing me to believe some sort of middleware design pattern is more suited, but also would like to avoid that.

At least for the time being, I think handling this on a provider-by-provider basis is okay!

See feedback though :)

fubuloubu commented on ape_alchemy/provider.py:
```diff
@@ -26,6 +27,12 @@ class Alchemy(Web3Provider, UpstreamProvider):
+        super().__init__(**kwargs)
+        self.block_page_size = 5000
+        self.concurrency = 1
+        print(f"setting block_page_size to {self.block_page_size} and concurrency to {self.concurrency}")
```
I don't think you need to log this, but definitely if you want to print anything, us loggers (as @unparalleled-js said)

wakamex replied:
ah yeah those were there temporarily! let me take them out!

fubuloubu reviewed (approved):
One comment, otherwise it looks good

wakamex commented:
incorporated the remaining comments

fubuloubu commented on ape_alchemy/provider.py:
```diff
@@ -132,23 +138,33 @@ def get_virtual_machine_error(self, exception: Exception) -> VirtualMachineError
+                )
+                if "exceeded its compute units" in message:
+                    retry_interval = min(max_retry_delay, min_retry_delay * retry_backoff_factor ** i)
+                    print(f"Alchemy compute units exceeded, retrying #{i} in {retry_interval} ms")
```
Would be great to convert this to an INFO log

wakamex replied:
just replaced it with logger.info() call. don't know if anything else is needed.

fubuloubu commented on ape_alchemy/provider.py:
```diff
@@ -132,23 +139,33 @@ def get_virtual_machine_error(self, exception: Exception) -> VirtualMachineError
-                else AlchemyProviderError
-            )
-            raise cls(message) from err
+    def _make_request(self, endpoint: str, parameters: list,min_retry_delay:int = 1000, retry_backoff_factor:int = 2,
+        max_retry_delay:int = 30000, max_retries:int = 3, retry_jitter:int = 250) -> Any:
```
```suggestion
    def _make_request(
        self,
        endpoint: str,
        parameters: list,
        min_retry_delay: int=1000,
        retry_backoff_factor: int=2,
        max_retry_delay: int=30000,
        max_retries: int=3,
        retry_jitter: int=250,
    ) -> Any:
```

fubuloubu reviewed (dismissed):
I think it might be best to get some of these parameters from the config object instead

perhaps @unparalleled-js can make a suggestion there

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -1,8 +1,10 @@
-import os
+import os, time
+import numpy.random as random
```
Is there any way we can use `random` instead? It is a built-in module to python. `numpy` is probably an indirect dependency of Ape and this plugin, however it is still protocol to include it in the `setup.py`... Thus, I think if we could use the built-in random module, it would be better. It has the same method that you use here `randint`

```suggestion
import random
```

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -26,6 +28,11 @@ class Alchemy(Web3Provider, UpstreamProvider):
 
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.block_page_size = 5000
```
Instead of setting these in an `init()`, set them as field values, like this:

```python
class Alchemy(Web3Provider, UpstreamProvider):
    ...

    block_page_size: int = 5000
    concurrency: int = 1
```

Maybe also include comments explaining why we need to override their defaults.
Then, no `init` overload is needed.

fubuloubu replied:
I was actually thinking these might be best driven via Config instead

wakamex replied:
> I was actually thinking these might be best driven via Config instead

check out this PR on ape https://github.com/ApeWorX/ape/pull/1157#discussion_r1033824686
that was my attempt to make these settings Config-based
i didn't know `ape --config` existed though, so i need to look into it more
judging by the existence of [ConfigWrapper](https://github.com/ApeWorX/ape/blob/45d47f9cc69a4065fc4fe15dd40c9c8cd836592e/src/ape/pytest/config.py#L9), it feels like both the command line and yaml feed into the same destination. adding ape-specific settings is needed in either case, which I think is @unparalleled-js's suggestion below. will look at that next.

fubuloubu replied:
Yeah, I think @unparalleled-js 's suggestion is ultimately the way to go. we allow plugins to define their own config and make it quite easy to use within the implementation, I think this is exactly the right use case for add'l config

unparalleled-js commented on ape_alchemy/provider.py:
```diff
@@ -132,23 +139,45 @@ def get_virtual_machine_error(self, exception: Exception) -> VirtualMachineError
+        endpoint: str,
+        parameters: list,
+        min_retry_delay: int=1_000, # 1 second
+        retry_backoff_factor: int=2, # exponential backoff
```
I think it is good to have these as kwargs, but it might also be nice to have them as properties you can set in the `ape-config` file.

The properties I mean are `min_retry_delay`, `retry_backoff_factor`, `max_retry_delay`, `max_retries`, and `retry_jitter`.

There currently is not a config class registered for this plugin, but I promise it's easy!
First, define the class at the top of the module:

```python
from ape.api import PluginConfig

class AlchemyConfig(PluginConfig):
   min_retry_delay: int = 1_000
   """
   The amount of milliseconds to wait before retrying the request.
   Defaults to one second.
   """
```

^ Put all of the same defaults you are using here in the config.

Then, in `__init__.py`, register it like this:

```python
from .provider import AlchemyConfig

@plugins.register(plugins.Config)
def config_class():
    yield AlchemyConfig
```

Then, in this method, all of these params can be `Optional[int]` and default to `None`. And you can grab each value like , taking in account of possible config values:

```python
from typing import cast

alchemy_config = cast(AlchemyConfig, self.config_manager.get_config("alchemy"))
retry_backoff_factor = retry_backoff_factor if retry_backoff_factor is not None else alchemy_config.retry_backoff_factor
...
```

fubuloubu commented:
Found this cool package today: https://pyratelimiter.readthedocs.io/en/latest/

gosuto-inzasheru commented:
love this! what is status? would hate to see this stale...

fubuloubu commented:
> love this! what is status? would hate to see this stale...

Quite stale, needs another champion u less @wakamex comes back and implements the feedback. Just a few things to add, and can also implement this for other providers

wakamex commented:
I learned how to use custom configs!

added `concurrency` and `block_page_size` to the custom config, so we can see nice outputs like this:
```py
for k,v in context.provider.config.__dict__.items():
    if not k.startswith('__'):
        print(f" {k} = {v}")
provider.config of type=<class 'ape_alchemy.provider.AlchemyConfig'>
 concurrency = 1
 block_page_size = 250000
 min_retry_delay = 1000
 retry_backoff_factor = 2
 max_retry_delay = 30000
 max_retries = 3
 retry_jitter = 250
```

I set `block_page_size` to a really big number, though I don't see any reason it shouldn't be infinite, since it seems to act as an upper limit. Here are my test results:
```
default:
    alchemy through ape took 78.44 seconds for 1291 events (events per second: 16.46)
block_page_size = 50_000:
    alchemy through ape took 0.85 seconds for 35 events (events per second: 41.09)
    alchemy through ape took 2.37 seconds for 1290 events (events per second: 544.34)
    alchemy through ape took 61.28 seconds for 40281 events (events per second: 657.27)
block_page_size = 10_000:
    alchemy through ape took 0.84 seconds for 35 events (events per second: 41.64)
    alchemy through ape took 1.72 seconds for 1290 events (events per second: 750.73)
    alchemy through ape took 29.22 seconds for 40281 events (events per second: 1378.50)
block_page_size = 50_000:
    alchemy through ape took 0.88 seconds for 35 events (events per second: 39.91)
    alchemy through ape took 1.09 seconds for 1290 events (events per second: 1179.53)
    alchemy through ape took 13.13 seconds for 40281 events (events per second: 3067.61)
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    INFO (ape-alchemy): Alchemy compute units exceeded, retrying, attempt 1/3 in 1000 ms
    alchemy through ape took 57.42 seconds for 123991 events (events per second: 2159.28)
block_page_size = 250_000:
    alchemy through ape took 0.64 seconds for 35 events (events per second: 54.67)
    alchemy through ape took 1.18 seconds for 1290 events (events per second: 1092.30)
    alchemy through ape took 12.72 seconds for 40281 events (events per second: 3165.94)
    alchemy through ape took 39.44 seconds for 123991 events (events per second: 3143.90)
infura default:
    infura through ape took 0.46 seconds for 35 events (events per second: 76.43)
    infura through ape took 1.01 seconds for 1290 events (events per second: 1276.06)
    infura through ape took 13.15 seconds for 40281 events (events per second: 3062.54)
    infura through ape took 41.21 seconds for 123991 events (events per second: 3008.86)
```

antazoey commented on ape_alchemy/provider.py:
```diff
@@ -34,15 +35,71 @@
+    """Configuration for Alchemy.
+
+    Fields
+    ------
```
we use google style docs!

antazoey commented on ape_alchemy/provider.py:
```diff
@@ -179,26 +236,60 @@ def get_virtual_machine_error(self, exception: Exception, **kwargs) -> VirtualMa
-                else AlchemyProviderError
-            )
-            raise cls(message) from err
+    def _make_request(
```
this is awesome but i don't think it'll apply to all the requests. Maybe that is ok for a first pass..

But, we need to use something web3.py middlerware or something such that all requests have rate-limiting handling logic.

antazoey commented on ape_alchemy/provider.py:
```diff
@@ -179,26 +236,60 @@ def get_virtual_machine_error(self, exception: Exception, **kwargs) -> VirtualMa
-                else AlchemyProviderError
-            )
-            raise cls(message) from err
+    def _make_request(
```
The base class has this same, internal method with different kwargs now, I wonder if that will cause problems?

wakamex replied:
if they're strictly after the existing (unchanged) signature, then I think it's probably fine? since they're strictly optional. baking them into the parameters parameter would clean this up, but I have no idea where that's created or what it's used for.

antazoey reviewed (commented):
the config is awesome and well thought out but as-is I don't think it applies to all requests made to Alchemy with this provider because not all requests use the internal `_make_request()` method: some use higher-level web3.py methods which have their own make_request within web3.py

wakamex commented:
removed class fields since every `Alchemy` class should have access to the `AlchemyConfig` I believe. Also removed `network_uris` since mypy didn't complain.

wakamex commented:
> not all requests use the internal `_make_request()` method: some use higher-level web3.py methods which have their own make_request within web3.py

which other methods need testing? I only tested this with `query`:
```py
curve_steth = Contract("0xDC24316b9AE028F1497c275EB9192a3Ea0f67022")
events = curve_steth.TokenExchange.query("*", start_block=chain.blocks[-1].number - int(86400*365/12))  # last 1 year
```

wakamex commented:
did you just merge your main into my main?

wakamex commented:
try me now
![image](https://github.com/ApeWorX/ape-alchemy/assets/16990562/5ab1363e-cc4e-452f-8f17-46a2e942c8c1)

fubuloubu commented:
Note: rate limit test would be hard and exhaust our already limited requests

antazoey commented:
> Note: rate limit test would be hard and exhaust our already limited requests

this implies you can't test Python logic without making live network requests.
I am not saying I am requiring tests, but I am calling this out as a false excuse.

wakamex commented:
with big `block_page_size` we can rest with 1 request that gets lots of
data. but since Alchemy uses "compute units" it'll still count lots toward
your rate limit. but I doubt you use your Alchemy API key a ton?

On Fri, Jan 26, 2024, 4:12 p.m. antazoey ***@***.***> wrote:

> Note: rate limit test would be hard and exhaust our already limited
> requests
>
> this implies you can't test Python logic without making live network
> requests.
> I am not saying I am requiring tests, but I am calling this out as a false
> excuse.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ApeWorX/ape-alchemy/pull/35#issuecomment-1912701836>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEBUCYR677IDESEFTQZMEB3YQQL2HAVCNFSM6AAAAAARQZE2OCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJSG4YDCOBTGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fubuloubu commented:
> > Note: rate limit test would be hard and exhaust our already limited requests
> 
> this implies you can't test Python logic without making live network requests.

That's fair, can unit test it with mocks and that works

fubuloubu commented:
> with big `block_page_size` we can rest with 1 request that gets lots of
> data. but since Alchemy uses "compute units" it'll still count lots toward
> your rate limit. but I doubt you use your Alchemy API key a ton?

We do use it quite a bit for some other plugins too, plus our own private testing use, so it's been a problem in the past

wakamex commented:
found a benchmark to justify infinite block page size

```py
connection = "alchemy through ape"
context = networks.parse_network_choice('ethereum:mainnet:alchemy')
context.__enter__();
legacy_abi = '[{"anonymous": false,"inputs": [{ "indexed": false, "name": "postTotalPooledEther", "type": "uint256" },{ "indexed": false, "name": "preTotalPooledEther", "type": "uint256" },{ "indexed": false, "name": "timeElapsed", "type": "uint256" },{ "indexed": false, "name": "totalShares", "type": "uint256" }],"name": "PostTotalShares","type": "event"}]'
oracle1 = Contract("0x442af784A788A5bd6F42A01Ebe9F287a871243fb", abi=legacy_abi)  # steth legacy oracle
v2_abi = '[{"anonymous": false,"inputs": [{ "indexed": true, "name": "reportTimestamp", "type": "uint256" },{ "indexed": false, "name": "timeElapsed", "type": "uint256" },{ "indexed": false, "name": "preTotalShares", "type": "uint256" },{ "indexed": false, "name": "preTotalEther", "type": "uint256" },{ "indexed": false, "name": "postTotalShares", "type": "uint256" },{ "indexed": false, "name": "postTotalEther", "type": "uint256" },{ "indexed": false, "name": "sharesMintedAsFees", "type": "uint256" }],"name": "TokenRebased","type": "event"}]'
steth = Contract("0xae7ab96520de3a18e5e111b5eaab095312d7fe84",abi=v2_abi)  # Lido v2 main contract

# %%
# local node through ape took 41.28 seconds for 1011 events (events per second: 24.49)
# 250k block page size: alchemy through ape took 10.33 seconds for 1012 events (events per second: 97.95)
# 2.5m block page size: alchemy through ape took 1.60 seconds for 1012 events (events per second: 631.47)
# 25m block page size: alchemy through ape took 0.90 seconds for 1012 events (events per second: 1128.32)
# infura through ape took 9.62 seconds for 1012 events (events per second: 105.17)
start_time = time.time()
events = oracle1.PostTotalShares.query("*")
print(f"{connection} took {time.time()-start_time:.2f} seconds for {len(events)} events (events per second: {len(events)/(time.time()-start_time):.2f})")

# %%
# local node through ape took 69.69 seconds for 267 events (events per second: 3.83)
# 250k block page size: alchemy through ape took 11.92 seconds for 268 events (events per second: 22.48)
# 2.5m block page size: alchemy through ape took 1.34 seconds for 268 events (events per second: 200.64)
# 25m block page size: alchemy through ape took 0.90 seconds for 1012 events (events per second: 1128.32)
# infura through ape took 9.39 seconds for 268 events (events per second: 28.55)
start_time = time.time()
events2 = steth.TokenRebased.query("*")
print(f"{connection} took {time.time()-start_time:.2f} seconds for {len(events2)} events (events per second: {len(events2)/(time.time()-start_time):.2f})")
```

antazoey commented:
> how would I mock up some tests? with like a MockLaggyProvider to connect to?

Using pytest.mock, you can basically mock anything in Python land.
Sometimes I mock web3 and then configure some side-effect responses which ca simulate waiting.

If it is too complex, I'd be happy to take a swing at it and share how I did it

wakamex commented:
I can take a look, pytest.mock sounds like it should be easy.

antazoey commented on ape_alchemy/provider.py:
```diff
@@ -179,26 +220,60 @@ def get_virtual_machine_error(self, exception: Exception, **kwargs) -> VirtualMa
+                message = str(err)
+                if any(
+                    error in message
+                    for error in ["exceeded its compute units", "Too Many Requests for url"]
```
I think this is the main part that changes per-provider (the specific error string to look for). I am thinking we may do what you suggested and move some of this to core when using default node connections (ape-geth), we can still receive this benefit.

Maybe we can a method to the base class that can be overriden like:

```python
def is_rate_limit_error(self, err):
```

fubuloubu commented on ape_alchemy/provider.py:
```diff
@@ -34,15 +36,57 @@
+            delay. Defaults to 250 milliseconds.
+    """
+
+    concurrency: int = 1  # can't do exponential backoff with multiple threads
```
Does this make sense to expose then?

wakamex commented:
started mocking up a test for the exponential backoff

fubuloubu commented:
this was the main inspiration for https://github.com/ApeWorX/ape/pull/2398

closing in favor of that impl
